boards: silabs: add support for EK2703A and minor doc fixes

vulpes2 opened the pull request:
1. Add support for xG24-EK2703A (EFR32xG24 Explorer Kit), based on existing files for DK2601B.
2. Add `EFR32MG24B210F1536IM48` SoC. The differences between `EFR32MG24B210F1536IM48` and `EFR32MG24B310F1536IM48` are not represented in device tree yet (high accuracy mode for ADC and GPIO pins)
3. SEGGER does not distribute firmware images for silabs J-Link OB adapters, remove the incorrect instructions and point users to Simplicity Studio instead.

asmellby commented on boards/silabs/dev_kits/xg24_ek2703a/Kconfig.xg24_ek2703a:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_XG24_EK2703A
+	select SOC_PART_NUMBER_EFR32MG24B310F1536IM48
```
The SoC OPN used on EK2703A is `EFR32MG24B210F1536IM48` (i.e. a 2 instead of a 3). It would need to be added to `dts/arm/silabs` and `soc/silabs`, as it isn't added in Zephyr yet. The difference between the two is that the 210 has 4 more GPIO pins, but does not have the high accuracy mode for IADC or the MVP accelerator. Neither of these are represented in device tree yet, so it makes no practical difference right now, but it would be good for the SoC selection to be correct in preparation for when they are.

vulpes2 replied:
Will do that along with the cleanup mentioned below in a separate PR.

asmellby commented on boards/silabs/dev_kits/xg24_ek2703a/Kconfig.defconfig:
```diff
@@ -0,0 +1,43 @@
+
+if BOARD_XG24_EK2703A
+
+config CMU_HFXO_FREQ
```
The HFXO and LFXO options no longer do anything, and should honestly be removed from all Series 2 boards, since this oscillators are now configured in device tree. However, this can be handled in a separate cleanup PR fixing all boards at once.

vulpes2 replied:
Noted, thanks.

asmellby commented on boards/silabs/dev_kits/xg24_ek2703a/board.cmake:
It would be nice to also support the Simplicity Commander runner (`silabs_commander`). See DK2602A for reference.

vulpes2 replied:
Noted, will investigate.

asmellby commented on boards/silabs/dev_kits/xg24_ek2703a/board.yml:
```diff
@@ -0,0 +1,6 @@
+  full_name: EFR32xG24 Explorer Kit (xG24-EK2703A)
+  vendor: silabs
+  socs:
+  - name: efr32mg24b310f1536im48
```
210

asmellby commented on boards/silabs/dev_kits/xg24_ek2703a/doc/index.rst:
```diff
@@ -0,0 +1,163 @@
+- CPU core: ARM Cortex®-M33 with FPU
+- Flash memory: 1536 kB
+- RAM: 256 kB
+- Transmit power: up to +20 dBm
```
210 is a 10 dBm OPN

asmellby commented on boards/silabs/dev_kits/xg24_ek2703a/doc/index.rst:
```diff
@@ -0,0 +1,163 @@
+
+The EFR32xG24 Explorer Kit (xG24-EK2703A) contains
+a Wireless System-On-Chip from the EFR32MG24 family built on an
+ARM Cortex®-M33F processor with excellent low power capabilities.
```
The core name is "M33", there should be no "F".

asmellby commented on boards/silabs/dev_kits/xg24_ek2703a/doc/index.rst:
```diff
@@ -0,0 +1,163 @@
+- RAM: 256 kB
+- Transmit power: up to +20 dBm
+- Operation frequency: 2.4 GHz
+- Crystals for LFXO (32.768 kHz) and HFXO (38.4 MHz).
```
The HFXO frequency on EK2703A is 39 MHz.

asmellby commented on boards/silabs/dev_kits/xg24_ek2703a/xg24_ek2703a.dts:
```diff
@@ -0,0 +1,206 @@
+ */
+
+/dts-v1/;
+#include <silabs/efr32mg24b310f1536im48.dtsi>
```
210

asmellby commented on boards/silabs/dev_kits/xg24_ek2703a/xg24_ek2703a_defconfig:
```diff
@@ -0,0 +1,12 @@
+CONFIG_GPIO=y
+CONFIG_HW_STACK_PROTECTION=y
+CONFIG_REGULATOR=y
+CONFIG_SOC_GECKO_EMU_DCDC=y
```
Like the oscillator options in the other defconfig file, these DCDC options don't do anything either, configuration is now done in device tree.

vulpes2 replied:
Deferred to a subsequent cleanup PR with the other defconfig file.

nordicjm commented on boards/silabs/dev_kits/xg24_ek2703a/Kconfig.defconfig:
```diff
@@ -0,0 +1,43 @@
+config CMU_LFXO_FREQ
+	default 32768
+
+config FLASH_BASE_ADDRESS
+	hex
+	default 0x08000000
```
this should come from dts

nordicjm commented on boards/silabs/dev_kits/xg24_ek2703a/board.c:
```diff
@@ -0,0 +1,35 @@
+
+	static struct gpio_dt_spec wake_up_gpio_dev =
+		GPIO_DT_SPEC_GET(DT_NODELABEL(wake_up_trigger), gpios);
+
+
```
double line

nordicjm commented on boards/silabs/dev_kits/xg24_ek2703a/doc/index.rst:
```diff
@@ -0,0 +1,166 @@
+Supported Features
+==================
+
+The board configuration supports the following hardware features:
```
Use board name as supplied to west, with \`\` around it and correct terminology https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#board-terminology

jhedberg commented on boards/silabs/dev_kits/xg24_ek2703a/Kconfig.defconfig:
```diff
@@ -0,0 +1,39 @@
+config COMMON_LIBC_MALLOC_ARENA_SIZE
+	default 8192
+
+config MAIN_STACK_SIZE
+	default 2304
```
I'm curious, where does this value come from? Isn't the main thread's stack requirement completely dependent on the application, i.e. why do we need to set some special default on the board level?

nordicjm replied:
Agreed, why are these stack sizes here?

vulpes2 replied:
This was a copy of the DK2601B config, I don't really understand why it has to be defined at the board level either, will remove it.

jhedberg commented on boards/silabs/dev_kits/xg24_ek2703a/Kconfig.defconfig:
```diff
@@ -0,0 +1,39 @@
+
+if SHELL
+
+config SHELL_STACK_SIZE
+	default 4096
```
Similar question here, i.e. doesn't this depend on which shell modules are enabled?

nordicjm commented on boards/silabs/dev_kits/xg24_ek2703a/Kconfig.defconfig:
```diff
@@ -0,0 +1,39 @@
+
+if BT
+
+config FPU
+	default y
```
duplicate

nordicjm commented on boards/silabs/dev_kits/xg24_ek2703a/Kconfig:
```diff
@@ -0,0 +1,12 @@
+
+if BOARD_XG24_EK2703A
+
+module = BOARD_EFR32MG24
```
also needs:
```
config BOARD_XG24_EK2703A
    select GPIO
    select BOARD_LATE_INIT_HOOK
```

nordicjm commented on boards/silabs/dev_kits/xg24_ek2703a/board.c:
```diff
@@ -0,0 +1,34 @@
+	return 0;
+}
+
+/* needs to be done after GPIO driver init */
+SYS_INIT(efr32xg24_ek2703a_init, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE);
```
remove

nordicjm commented on boards/silabs/dev_kits/xg24_ek2703a/board.c:
```diff
@@ -0,0 +1,34 @@
+
+LOG_MODULE_REGISTER(efr32xg24_ek2703a, CONFIG_BOARD_EFR32MG24_LOG_LEVEL);
+
+static int efr32xg24_ek2703a_init(void)
```
and replace with `void board_late_init_hook(void)`

nordicjm commented on boards/silabs/dev_kits/xg24_ek2703a/doc/index.rst:
```diff
@@ -0,0 +1,167 @@
+Supported Features
+==================
+
+The ``xg24_ek2703a`` board configuration supports the following hardware
```
-configuration

nordicjm commented on boards/silabs/dev_kits/xg24_ek2703a/xg24_ek2703a.dts:
```diff
@@ -0,0 +1,206 @@
+			reg = <0x00080000 0x00074000>;
+		};
+
+		/* Reserve 32 kB for the scratch partition */
+		scratch_partition: partition@f4000 {
+			label = "image-scratch";
+			reg = <0x000f4000 0x00008000>;
+		};
```
scratch really shouldn't be used

nordicjm commented on soc/silabs/silabs_s2/efr32mg24/Kconfig.soc:
```diff
@@ -19,15 +19,21 @@ config SOC_PART_NUMBER_EFR32MG24B310F1536IM48
+config SOC_PART_NUMBER_EFR32MG24B210F1536IM48
+	bool
+	select SOC_SERIES_EFR32MG24
+
```
needs splitting into 2 or 3 commits, one adding the soc/dts, the other adding the board

vulpes2 replied:
Done

nordicjm commented on boards/silabs/dev_kits/xg24_ek2703a/board.cmake:
```diff
@@ -3,3 +3,6 @@
 board_runner_args(jlink "--device=EFR32MG24BxxxF1536" "--reset-after-load")
 include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
+
+board_runner_args(silabs_commander "--device=EFR32MG24B210F1536IM48")
+include(${ZEPHYR_BASE}/boards/common/silabs_commander.board.cmake)
```
move this into the board commit, doesn't make sense to add the board in one commit then edit it right away in the following commit

nordicjm commented on boards/silabs/dev_kits/xg24_ek2703a/doc/index.rst:
```diff
@@ -102,7 +102,7 @@ Programming and Debugging
 .. note::
    Before using the kit the first time, you should update the J-Link firmware
-   from `J-Link-Downloads`_
+   in Simplicity Studio.
```
likewise, add this to the board commit

vulpes2 committed:
doc: silabs: fix J-Link firmware update instructions

J-Link firmware images for silabs boards are not distributed by
SEGGER, fw update should be done with Simplicity Studio instead.

Signed-off-by: Daniel Fuchs <software@sagacioussuricata.com>

vulpes2 committed:
soc: introduce the EFR32MG24B210F1536IM48

The EFR32MG24B210F1536IM48 has 4 more GPIOs than the
EFR32MG24B310F1536IM48, and does not support the high accuracy
mode for the IADC.

Signed-off-by: Daniel Fuchs <software@sagacioussuricata.com>

vulpes2 committed:
boards: silabs: add EK2703A support

Based on board files for DK2601B. The EFR32xG24 Explorer Kit
(xG24-EK2703A) is similar to the DK2601B but does not have any
sensors onboard.

Signed-off-by: Daniel Fuchs <software@sagacioussuricata.com>
